lib/flop_phoenix/form_data.ex

```diff
@@ -0,0 +1,143 @@
+defimpl Phoenix.HTML.FormData, for: Flop.Meta do
+  import Flop.Phoenix.Misc, only: [maybe_put: 3]
+
+  def to_form(meta, opts) do
+    {name, opts} = name_and_opts(meta, opts)
+    {errors, opts} = Keyword.pop(opts, :errors, [])
+    {hidden, opts} = Keyword.pop(opts, :hidden, [])
+    {params, opts} = Keyword.pop(opts, :params, %{})
+    id = Keyword.get(opts, :id) || name || "flop"
+
+    %Phoenix.HTML.Form{
+      data: meta.flop,
+      errors: errors,
+      hidden: hidden_inputs(meta.flop, hidden),
+      id: id,
+      impl: __MODULE__,
+      name: name,
+      options: opts,
+      params: params,
+      source: meta
+    }
+  end
+
+  def to_form(
+        meta,
+        %{data: %Flop{} = flop} = form,
+        :filters,
+        opts
+      ) do
+    no_unsupported_options!(opts)
+
+    {id, opts} = Keyword.pop(opts, :id)
+    {default, opts} = Keyword.pop(opts, :default, [])
+    {prepend, opts} = Keyword.pop(opts, :prepend, [])
+    {append, opts} = Keyword.pop(opts, :append, [])
+    {skip_hidden_op, opts} = Keyword.pop(opts, :skip_hidden_op, false)
+
+    name = if form.name, do: form.name <> "[filters]", else: "filters"
+    id = if id = id || form.id, do: to_string(id <> "_filters"), else: "filters"
+    filters = if flop.filters == [], do: default, else: flop.filters
```

Comment:
🤔 

Assuming `default` is where `filter_inputs_for` would pass in the fields as given in its `fields` assign, IMO we should make the resulting filters always match these, as otherwise removing parts of the `&filters[]...` query params from the URL would result in filter fields not being rendered. So something like


```
filters = Enum.map(default, &merge_op_and_value_from_filter(&1, flop.filters))
```

iow the given `fields` configuration should always dictate the form

Reply:
hmm possibly I haven't fully understood how the `FormData` protocol works, but I guess you get what I mean.

Possibly in the `input_value` below we should instead generate a `Flop.Filter` for each filter from the default, regardless of whether it was present in the current request

Reply:
The `:default` option is described in the docs of `Phoenix.HTML.Form.inputs_for`, so we shouldn't deviate from that. But we process any number of additional opts. So we can add a `:fields` opt here to always render a fixed layout regardless of the content of `Flop.Filters`.